controllers/topology/current_state.go

```diff
@@ -18,12 +18,132 @@ package topology
 
 import (
 	"context"
+	"fmt"
 
-	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/cluster-api/api/v1alpha4"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-// Gets the current state of the Cluster topology.
-func (r *ClusterReconciler) getCurrentState(ctx context.Context, cluster *clusterv1.Cluster) (*clusterTopologyState, error) {
-	// TODO: add get class logic; also remove nolint exception from clusterTopologyState and machineDeploymentTopologyState
-	return nil, nil
+// getCurrentState gets information about the current state of a Cluster by inspecting the state of the InfrastructureCluster,
+// the ControlPlane, and the MachineDeployments associated with the Cluster.
+func (r *ClusterReconciler) getCurrentState(ctx context.Context, cluster *v1alpha4.Cluster, class v1alpha4.ClusterClass) (*clusterTopologyState, error) {
+	clusterState := clusterTopologyState{
+		cluster: cluster,
+	}
+
+	// Reference to the InfrastructureCluster can be nil and is expected to be on the first reconcile.
+	// In this case the method should still be allowed to continue.
+	if cluster.Spec.InfrastructureRef != nil {
+		infra, err := r.getCurrentInfrastructureClusterState(ctx, cluster)
+		if err != nil {
+			return nil, err
+		}
+		clusterState.infrastructureCluster = infra
+	}
+
+	// Reference to the ControlPlane can be nil, and is expected to be on the first reconcile. In this case the method
+	// should still be allowed to continue.
+	if cluster.Spec.ControlPlaneRef != nil {
+		cp, err := r.getCurrentControlPlaneState(ctx, cluster, class)
+		if err != nil {
+			return nil, err
+		}
+		clusterState.controlPlane = cp
+	}
+
+	// A Cluster may have zero or more MachineDeployments and a Cluster is expected to have zero MachineDeployments on
+	// first reconcile.
+	m, err := r.getCurrentMachineDeploymentState(ctx, cluster)
+	if err != nil {
+		return nil, err
+	}
+	clusterState.machineDeployments = m
+	return &clusterState, nil
+}
+
+// getCurrentInfrastructureClusterState looks for the state of the InfrastructureCluster. If a reference is set but not
+// found, either from an error or the object not being found, an error is thrown.
+func (r *ClusterReconciler) getCurrentInfrastructureClusterState(ctx context.Context, cluster *v1alpha4.Cluster) (*unstructured.Unstructured, error) {
+	infra, err := r.getTemplate(ctx, cluster.Spec.InfrastructureRef)
```

Comment:
This was previously renamed to getReference in this PR (afaik), was this an intentional rollback during rebase?

Reply:
This was an error on my rebase. I'll change it back to getReference.